atlasdb-ete-tests/src/test/java/com/palantir/atlasdb/ete/QosCassandraWriteEteTest.java

```diff
@@ -0,0 +1,123 @@
+ *
+ * http://opensource.org/licenses/BSD-3-Clause
+ *
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS IS" BASIS,
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+
+package com.palantir.atlasdb.ete;
+
+import static org.assertj.core.api.Assertions.assertThat;
+import static org.assertj.core.api.Assertions.assertThatThrownBy;
+
+import java.util.ArrayList;
+import java.util.List;
+import java.util.concurrent.ExecutionException;
+import java.util.concurrent.ExecutorService;
+import java.util.concurrent.Executors;
+import java.util.concurrent.Future;
+import java.util.concurrent.TimeUnit;
+import java.util.concurrent.atomic.AtomicInteger;
+import java.util.stream.IntStream;
+
+import org.junit.Test;
+
+import com.google.common.base.Preconditions;
+import com.google.common.base.Stopwatch;
+import com.google.common.base.Throwables;
+import com.palantir.atlasdb.qos.ratelimit.RateLimitExceededException;
+
+public class QosCassandraWriteEteTest extends QosCassandraEteTestSetup {
+
+    @Test
+    public void shouldBeAbleToWriteSmallAmountOfBytesIfDoesNotExceedLimit() {
+        writeNTodosOfSize(1, 100);
+    }
+
+    @Test
+    public void shouldBeAbleToWriteSmallAmountOfBytesSeriallyIfDoesNotExceedLimit() {
+        IntStream.range(0, 50).forEach(i -> writeNTodosOfSize(1, 100));
+    }
+
+    @Test
+    public void shouldBeAbleToWriteLargeAmountsExceedingTheLimitFirstTime() {
+        writeNTodosOfSize(12, 1_000);
+    }
+
+    @Test
+    public void shouldBeAbleToWriteLargeAmountsExceedingTheLimitSecondTimeWithSoftLimiting() {
+        Stopwatch stopwatch = Stopwatch.createStarted();
+        writeNTodosOfSize(1, 20_000);
+        long firstWriteTime = stopwatch.elapsed(TimeUnit.MILLISECONDS);
+
+        stopwatch = Stopwatch.createStarted();
+        writeNTodosOfSize(200, 1_000);
+        long secondWriteTime = stopwatch.elapsed(TimeUnit.MILLISECONDS);
+
+        assertThat(secondWriteTime).isGreaterThan(firstWriteTime);
+        assertThat(secondWriteTime - firstWriteTime).isLessThan(MAX_SOFT_LIMITING_SLEEP_MILLIS);
+    }
+
+    @Test
+    public void shouldNotBeAbleToWriteLargeAmountsIfSoftLimitSleepWillBeMoreThanConfiguredBackoffTime() {
+        // Have one quick limit-exceeding write, as the rate-limiter
+        // will let anything pass through until the limit is exceeded.
+        writeNTodosOfSize(1, 100_000);
+
+        assertThatThrownBy(() -> writeNTodosOfSize(200, 10_000))
+                .isInstanceOf(RateLimitExceededException.class)
+                .hasMessage("Rate limited. Available capacity has been exhausted.");
+    }
+
+    @Test
+    public void writeRateLimitShouldBeRespectedByConcurrentWritingThreads() throws InterruptedException {
+        int oneTodoSizeInBytes = 167;
```

Comment:
nit: was hoping to have a programmatic way of generating this, though definitely non-blocking

Reply:
I thought about that, but its not clean enough, it depends on the size of the encoded row, not easy to calculate directly, but we might be able to leverage `ThriftObjectSizeUtils` but that kind of defeats the purpose of a test, as we start using implementation details.